<?xml version="1.6" ?>
<launch>

    <!-- IP Values -->
    <arg name="rasp1_IP" default="192.168.10.101" />
    <arg name="rasp2_IP" default="192.168.10.102" />
    <arg name="rasp3_IP" default="192.168.10.103" />
    <arg name="rasp4_IP" default="192.168.10.104" />

    <arg name="central_IP" default="$(optenv ROS_MASTER_IP 192.168.10.105)" />

    <arg name="back_Lidar_IP" default="192.168.10.111"/>
	<arg name="front_Lidar_IP" default="192.168.10.112"/>


    <!--Node Names -->
    <arg name="right_motor_node_name" default="right_side_wheels" />
    <arg name="left_motor_node_name" default="left_side_wheels" />

	<arg name="back_sick_node_name" default="sick_back_rpi_3"/>
	<arg name="front_sick_node_name" default="sick_front_rpi_4"/>

	<arg name="back_findlines_node_name" default="engrais_findlines_back"/>
	<arg name="front_findlines_node_name" default="engrais_findlines_front"/>

    <arg name="central_node_name" default="engrais_central"/>
    <arg name="controller_move_node_name" default="controller_move_node" />


    <!-- ROS Topics -->
    <arg name="right_wheel_control_signal" default="rightCommand" />
    <arg name="left_wheel_control_signal" default="leftCommand" />
    <arg name="emer_topic_name" default="emergency_topic"/>

	<arg name="back_scan_topic" default="scan_back"/>
	<arg name="front_scan_topic" default="scan_front"/>

	<arg name="back_rviz_frame" default="laser_back"/>
	<arg name="front_rviz_frame" default="laser_front"/>

    <arg name="change_mode_topic" default="mode_control"/>
    <arg name="central_selected_lines_topic" default="selectedLines"/>


	<!-- Rviz Frame Names -->
	<arg name="back_lines_topic" default="lines_back"/>
	<arg name="front_lines_topic" default="lines_front"/>
	
	<arg name="central_global_frame" default="body"/>


    <!-- USB Names -->
    <arg name="right_serial_back_port" default="/dev/ttyUSB0" />
    <arg name="right_serial_front_port" default="/dev/ttyUSB1" />

    <arg name="left_serial_back_port" default="/dev/ttyUSB0" />
    <arg name="left_serial_front_port" default="/dev/ttyUSB1" />


    <!-- Robot Infos -->
	<arg name="bluetooth_device_address" default="00:04:4b:a5:db:33" />
	<arg name="device_name" default="Wireless Controller" />
	

    <!-- Controller Infos -->
	<arg name="right_robot_side" default="right" />
	<arg name="left_robot_side" default="left" />


	<!-- Robot's Infos -->
    <arg name="mode" default="manual"/>
	<arg name="max_velocity" default="4.0"/>
	<arg name="body_size" default="2.0"/>
	<arg name="distance_reference" default="1.5"/>


	<!-- Common Serial Arguments -->
    <arg name="serial_baud_value" default="57600" />
    <arg name="serial_parity_value" default="0" /> <!-- 0:None, 1:odd, 2:even -->
    <arg name="serial_stop_bit_value" default="1" /> <!-- 1:One, 2:Two, 3:One point five -->
    <arg name="serial_data_bits_value" default="8" /> <!-- 5, 6, 7 or 8 -->
    <arg name="serial_hdw_flow_ctrl_value" default="0" /> <!-- 0:None, 1:software, 2:hardware -->
    <arg name="serial_timeout_value" default="10" />


	<!-- Control Period -->
	<arg name="sleep_time_ms" default="500"/>


    <!-- Connect to Rasp1 -->
 	<machine name="rasp1" address="$(arg rasp1_IP)" user="engrais" password="engrais" env-loader="~/catkin_ws/src/istiaENGRAIS/config/config.sh" default="true"/>

 	<include file="$(find engrais_motors)/launch/motor.launch">
	    <arg name="motor_node_name" value="$(arg right_motor_node_name)" />
	    <arg name="sub_topic_name" value="$(arg right_wheel_control_signal)" />
        <arg name="emer_topic" value="$(arg emer_topic_name)"/>

	    <arg name="back_port_name" value="$(arg right_serial_back_port)" />
	    <arg name="front_port_name" value="$(arg right_serial_front_port)" />

	    <arg name="robot_side" value="$(arg right_robot_side)" />

	    <arg name="serial_baud" value="$(arg serial_baud_value)" />
	    <arg name="serial_parity" value="$(arg serial_parity_value)" />
	    <arg name="serial_stop_bit" value="$(arg serial_stop_bit_value)" />
	    <arg name="serial_data_bits" value="$(arg serial_data_bits_value)" />
	    <arg name="serial_hdw_flow_ctrl" value="$(arg serial_hdw_flow_ctrl_value)" />
	    <arg name="serial_timeout" value="$(arg serial_timeout_value)" />
 	</include>


 	<!-- Connect to Rasp2 -->
 	<machine name="rasp2" address="$(arg rasp2_IP)" user="engrais" password="engrais" env-loader="~/catkin_ws/src/istiaENGRAIS/config/config.sh" default="true"/>

 	<include file="$(find engrais_motors)/launch/motor.launch">
	    <arg name="motor_node_name" value="$(arg left_motor_node_name)" />
	    <arg name="sub_topic_name" value="$(arg left_wheel_control_signal)" />
        <arg name="emer_topic" value="$(arg emer_topic_name)"/>

	    <arg name="back_port_name" value="$(arg left_serial_back_port)" />
	    <arg name="front_port_name" value="$(arg left_serial_front_port)" />

	    <arg name="robot_side" value="$(arg left_robot_side)" />

	    <arg name="serial_baud" value="$(arg serial_baud_value)" />
	    <arg name="serial_parity" value="$(arg serial_parity_value)" />
	    <arg name="serial_stop_bit" value="$(arg serial_stop_bit_value)" />
	    <arg name="serial_data_bits" value="$(arg serial_data_bits_value)" />
	    <arg name="serial_hdw_flow_ctrl" value="$(arg serial_hdw_flow_ctrl_value)" />
	    <arg name="serial_timeout" value="$(arg serial_timeout_value)" />
 	</include>


	<!-- Connect to Rasp3 -->
 	<machine name="rasp3" address="$(arg rasp3_IP)" user="engrais" password="engrais" env-loader="~/catkin_ws/src/istiaENGRAIS/config/config.sh" default="true"/>

 	<include file="$(find engrais_control)/launch/findlines.launch">
	    <arg name="IP_Lidar" value="$(arg front_Lidar_IP)" />
	    <arg name="sick_node_name" value="$(arg front_sick_node_name)" />
	    <arg name="scan_topic" value="$(arg front_scan_topic)" />
	    <arg name="rviz_frame" value="$(arg front_rviz_frame)" />

	    <arg name="lines_topic" value="$(arg front_lines_topic)" />
	    <arg name="findlines_node_name" value="$(arg front_findlines_node_name)" />
        <arg name="emer_topic" value="$(arg emer_topic_name)"/>
 	</include>


 	<!-- Connect to Rasp4 -->
 	<machine name="rasp4" address="$(arg rasp4_IP)" user="engrais" password="engrais" env-loader="~/catkin_ws/src/istiaENGRAIS/config/config.sh" default="true"/>

 	<include file="$(find engrais_control)/launch/findlines.launch">
	    <arg name="IP_Lidar" value="$(arg back_Lidar_IP)" />
	    <arg name="sick_node_name" value="$(arg back_sick_node_name)" />
	    <arg name="scan_topic" value="$(arg back_scan_topic)" />
	    <arg name="rviz_frame" value="$(arg back_rviz_frame)" />

	    <arg name="lines_topic" value="$(arg back_lines_topic)" />
	    <arg name="findlines_node_name" value="$(arg back_findlines_node_name)" />
        <arg name="emer_topic" value="$(arg emer_topic_name)"/>
 	</include>


 	<!-- Connect to Central -->
 	<machine name="central" address="$(arg central_IP)" user="jetson-tx2" password="jetsontx2" env-loader="~/catkin_ws/src/istiaENGRAIS/config/config.sh" default="true"/>

 	<include file="$(find engrais_control)/launch/control.launch">
        <arg name="control_node_name" value="$(arg central_node_name)"/>

		<arg name="robot_max_velocity" value="$(arg max_velocity)"/>
		<arg name="robot_body_size" value="$(arg body_size)"/>
		<arg name="robot_distance_reference" value="$(arg distance_reference)"/>
        <arg name="control_sleep_time_ms" value="$(arg sleep_time_ms)"/>

        <arg name="emer_topic" value="$(arg emer_topic_name)"/>
		<arg name="change_mode_topic_name" value="$(arg change_mode_topic)"/>
		
		<arg name="subscribe_topic_back" value="$(arg back_lines_topic)"/>
		<arg name="subscribe_topic_front" value="$(arg front_lines_topic)"/>
		<arg name="publish_topic_left" value="$(arg left_wheel_control_signal)"/>
		<arg name="publish_topic_right" value="$(arg right_wheel_control_signal)"/>

		<arg name="global_frame" value="$(arg central_global_frame)"/>
        <arg name="robot_mode" value="$(arg mode)"/>
		<arg name="selected_lines_topic" value="$(arg central_selected_lines_topic)"/>
 	</include>


    <node name="$(arg controller_move_node_name)" pkg="engrais" type="robot_controller_move.py" output="screen">
        <param name="bluetooth_address" type="string" value="$(arg bluetooth_device_address)"/>        
        <param name="bluetooth_device_name" type="string" value="$(arg device_name)"/>
        
        <param name="pub_topic_left" type="string" value="$(arg left_wheel_control_signal)"/>        
        <param name="pub_topic_right" type="string" value="$(arg right_wheel_control_signal)"/>
        
        <param name="mode_change_topic" type="string" value="$(arg change_mode_topic)"/>
        <param name="emergency_topic" type="string" value="$(arg emer_topic_name)"/>

        <param name="desired_velocity" type="double" value="$(arg max_velocity)"/>
    </node>

</launch>
